### PR TITLE
docs(README): add cross and multimodal documentation into readme

### PR DESCRIPTION
**Pull request by JoanFM**
**Changes introduced**
Introducing Cross and MultiModality documentation in the README

### CODE REVIEW

**Review by alexcg1 — COMMENTED** [review]
I'm concerned that our README is already quite overloaded with content - I think a good way moving forwards would be to have this as a doc that we can link to from the Learn section of readme

**Review by hanxiao — CHANGES_REQUESTED** [review]
It's not about overloaded, I think this added section is too wordy and does not fit the README style we have. Principles:
- lean code first
- text only if necessary

**Comment by JoanFM** [comment]
> I think this added section is too wordy and does not fit to the README style we have. Principles:
> 
> * lean code first
> * text only if necessary

Okey, perfect, will try to adapt it

File: README.md [hunk 1]
```diff
@@ -354,6 +354,51 @@ pip install jina[hub] && jina hub new --type app
...
+However, this is not accurate. For example, one can do cross-modal search by searching images from different points of view, 
+or searching for matching titles for given paragraph text.
+Therefore, one can consider that a modality is related to a given data distribution from which input may come. 
+
+For this reason, and to have first-class support for cross and multi-modal search, Jina offers modality as an attribute in its Document primitive type.
+Now that we are agreed on the concept of modality, we can describe cross-modal and multi-modal search.
+ 
+ - Cross-modal search can be defined as a set of retrieval applications that try to effectively find relevant documents of modality A by querying with documents from modality B.
+ - Multi-modal search can be defined as a set of retrieval applications that try to effectively project documents of different modalities into a
+ common embedding space.
```
Comment (bwanglzu) [inline README.md:375]: into a common embedding space and find relevant document with respect to the fusion of multiple modalities

File: README.md [hunk 2]
```diff
@@ -354,6 +354,51 @@ pip install jina[hub] && jina hub new --type app
...
+For this reason, and to have first-class support for cross and multi-modal search, Jina offers modality as an attribute in its Document primitive type.
+Now that we are agreed on the concept of modality, we can describe cross-modal and multi-modal search.
+ 
+ - Cross-modal search can be defined as a set of retrieval applications that try to effectively find relevant documents of modality A by querying with documents from modality B.
+ - Multi-modal search can be defined as a set of retrieval applications that try to effectively project documents of different modalities into a
+ common embedding space. 
+ 
+The main difference between these two search modes is that for cross-modal, there is a direct mapping between a single document or chunk and a
+vector in embedding space, while for MultiModal this does not hold true, since 2 or more documents might be combined into a single vector.
+
```
Comment (bwanglzu) [inline README.md:379]: this does not hold true -> does not hold ~true~

Comment (alexcg1) [reply README.md:379]: does not hold true is decent English ;)

Comment (bwanglzu) [reply README.md:379]: ok good to know haha

File: README.md [hunk 3]
```diff
@@ -354,6 +354,51 @@ pip install jina[hub] && jina hub new --type app
 
 This creates a Python entrypoint, YAML configs and a Dockerfile. You can start from there.
 
+## Multi & Cross Modality
+
+Jina is a data type agnostic framework. It basically can work with any data type.
```
Comment (cristianmtr) [inline README.md:359]:
```suggestion
Jina is a data type-agnostic framework. It basically can work with any data type. 
```

File: README.md [hunk 4]
```diff
@@ -354,6 +354,51 @@ pip install jina[hub] && jina hub new --type app
 
 This creates a Python entrypoint, YAML configs and a Dockerfile. You can start from there.
 
+## Multi & Cross Modality
+
+Jina is a data type agnostic framework. It basically can work with any data type. 
+One just needs to find a way to project the documents of this type into an embedding space.
+
+But Jina does not only offer this but also allows you to run cross and multi-modal search flows.
```
Comment (cristianmtr) [inline README.md:362]:
```suggestion
But Jina does not only offer this, but also allows you to run cross- and multi-modal search flows. 
```

Comment (nan-wang) [reply README.md:362]: `But ... but also` sounds wired.

**Review by nan-wang — CHANGES_REQUESTED** [review]
The story needs to be further touched. I think a better structure might be 
1. introduce the concept of modality
2. introduce cross-modality
3. introduce multi-modality

File: README.md [hunk 5]
```diff
@@ -354,6 +354,51 @@ pip install jina[hub] && jina hub new --type app
 
 This creates a Python entrypoint, YAML configs and a Dockerfile. You can start from there.
 
+## Multi & Cross Modality
+
+Jina is a data type agnostic framework. It basically can work with any data type. 
+One just needs to find a way to project the documents of this type into an embedding space.
```
Comment (nan-wang) [inline README.md:360]: This part might digress a little from the topic of modality.

File: README.md [hunk 6]
```diff
@@ -354,6 +354,51 @@ pip install jina[hub] && jina hub new --type app
...
+### Cross modal search
+
+Supporting cross-modal search in Jina is very easy. One just needs to properly set the modality tags of the input documents
+and design the Flow in such a way that the queries target the desired embedding space.
+
+### Multi modal search
+
+In order to support MultiModal search and to make it easy to build such applications, jina provides three new concepts.
+ 
+ - MultiModalDocument: MultiModalDocument is a document composed by more than one chunk with different modalities. It makes it easy
```
Comment (nan-wang) [inline README.md:394]:
```suggestion
 - `MultiModalDocument`: `MultiModalDocument` is a document composed of more than one chunk with different modalities. It makes it easy
```

**Review by nan-wang — CHANGES_REQUESTED** [review]

**Comment by hanxiao** [comment]
> The story needs to be further touched. I think a better structure might be 
> 
> 1. introduce the concept of modality
> 
> 2. introduce cross-modality
> 
> 3. introduce multi-modality
> 
> 

please don't. see my last comment, this readme is too much word.

lean usage code only, text only when necessary

**Comment by hanxiao** [comment]
i'm closing this PR and everybody who commented on this PR please first get a feeling on our current README writing style, and understand what do I mean by "lean usage code first, text only when necessary".

File: README.md (hunk 3 above)
Comment (JoanFM) [reply README.md:359]: are we sure? data type go together, it seems weird. @alexcg1 what would be correct English?

Comment (cristianmtr) [reply README.md:359]: image
There are certain words, that, when used as suffixes, require a dash (`-`)

Comment (cristianmtr) [reply README.md:359]: image

It's a v minor thing :P

File: README.md [hunk 7]
```diff
@@ -394,6 +394,113 @@ This creates a Python entrypoint, YAML configs and a Dockerfile. You can start f
 
 </details>
 
+#### MultiModalDocument
```
Comment (hanxiao) [inline README.md:397]:
```suggestion
#### MultimodalDocument
```

File: README.md [hunk 8]
```diff
@@ -394,6 +394,113 @@ This creates a Python entrypoint, YAML configs and a Dockerfile. You can start f
...
+    chunk_image.blob =  Image.open('path/to/image.jpg')
+    chunk_image.modality = 'description'
+    chunk_image.tags['date'] = '10/08/2019' 
+    
+    document = Document()
+    document.chunks.append(chunk_title)
+    document.chunks.append(chunk_description)
+    document.chunks.append(chunk_image)
+
+    multimodal_document = MultimodalDocument(document)
```
Comment (hanxiao) [inline README.md:470]: too long

File: README.md (hunk 3 above)
Comment (JoanFM) [reply README.md]: Yes, my point is then datatype belongs together?

File: README.md (hunk 4 above)
Comment (JoanFM) [reply README.md]: Done!

File: README.md [hunk 9]
```diff
@@ -394,6 +394,79 @@ This creates a Python entrypoint, YAML configs and a Dockerfile. You can start f
 
 </details>
 
+#### MultimodalDocument
+  
+ A MultimodalDocument in Jina is a document composed by more than one chunks with different modalities.
```
Comment (yuanbit) [inline README.md:399]: chunk

File: README.md [hunk 10]
```diff
@@ -394,6 +394,79 @@ This creates a Python entrypoint, YAML configs and a Dockerfile. You can start f
...
+```python
+    from jina.types import Document
+    from jina.types.document.multimodal import MultimodalDocument
+    chunk_title = Document('my holiday picture')
+    chunk_title.modality = 'title'
+    
+    chunk_description = Document('the family having fun on the beach')
+    chunk_description.modality = 'description'
+
+    chunk_image = Document(Image.open('path/to/image.jpg'))
```
Comment (yuanbit) [inline README.md:431]: I think the "From content modality mapping" and "From chunks with different modalities" are not necessary. To reduce the number of words, I wonder if we could just say:

We can build a multimodal document by providing the modality names and the content of the chunks:

`Code`

We can also create chunks when we have extra metadata in each chunk:

`Code`

File: README.md [hunk 11]
```diff
@@ -394,6 +394,79 @@ This creates a Python entrypoint, YAML configs and a Dockerfile. You can start f
...
+    chunk_image.modality = 'description'
+    chunk_image.tags['date'] = '10/08/2019' 
+  
+    document = MultimodalDocument(chunks=[chunk_title, chunk_description, chunk_image])
+```
+
+#### MultimodalEncoder and MultimodalDriver
+
+In order to extract fusion embeddings from different modalities we have `MultimodalEncoders`.
```
Comment (yuanbit) [inline README.md:440]: I think "fusion embeddings" is not so clear for someone new to multimodal search. Perhaps we can say:

"We can use [MultimodalEncoders](https://github.com/jina-ai/examples/tree/master/multimodal-search-tirg) to get the feature embeddings from different modalities."

Comment (yuanbit) [inline README.md:440]: I think "fusion embeddings" is not so clear for someone new to multimodal search. Perhaps we can say:

"We can use [MultimodalEncoders](https://github.com/jina-ai/examples/tree/master/multimodal-search-tirg) to get the feature embeddings from different modalities."

File: README.md [hunk 12]
```diff
@@ -394,6 +394,79 @@ This creates a Python entrypoint, YAML configs and a Dockerfile. You can start f
...
+  texts_path: texts.pkl
+  positional_modality: ['image', 'text']
+requests:
+  on:
+    [IndexRequest, SearchRequest]:
+      - !MultiModalDriver {}
+```
+
+A `MultimodalEncoder` expects a variable number of arguments in the `encode` interface.
+
```
Comment (yuanbit) [inline README.md:456]: I think it is enough to have this in the documentation and not in the readme.

File: docs/chapters/cross-multi-modality/index.md [hunk 13]
```diff
@@ -0,0 +1,42 @@
+## Multi & Cross Modality
+
+Jina is a data type-agnostic framework. It basically can work with any kind of data.
```
Comment (yuanbit) [inline docs/chapters/cross-multi-modality/index.md:3]: "Jina is a data type-agnostic framework and can work with any kind of data."

File: README.md (hunk 12 above)
Comment (yuanbit) [reply README.md:456]: Perhaps we can say "Learn more" and link it to the documentation page.

File: README.md (hunk 9 above)
Comment (yuanbit) [inline README.md:399]: I wonder if we should use "input" instead of "document" for people not familiar with Jina terminologies.

File: docs/chapters/cross-multi-modality/index.md (hunk 13 above)
Comment (yuanbit) [inline docs/chapters/cross-multi-modality/index.md:3]: "Jina is a data type-agnostic framework, offering one to work with any type of data and run cross- and multi-modal search flows."

File: docs/chapters/cross-multi-modality/index.md [hunk 14]
```diff
@@ -0,0 +1,42 @@
+## Multi & Cross Modality
+
+Jina is a data type-agnostic framework. It basically can work with any kind of data. 
+
+Jina does not only offer the possibility to work with any data, but also allows you to run cross- and multi-modal search flows. 
+To better understand what this implies we first need to understand the concept of modality. 
+
+One may think that different modalities correspond to different kinds of data (images and text in this case).
+However, this is not accurate. For example, one can do cross-modal search by searching images from different points of view,
```
Comment (yuanbit) [inline docs/chapters/cross-multi-modality/index.md:9]: "A misconception is that modalities correspond to different kinds of data, such as images or text. However, one can also do a cross-modal search of images from images of different points of view or paragraphs of text from matching titles. Therefore, one can think of a modality as a given data distribution of the input."

File: docs/chapters/cross-multi-modality/index.md [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+ - Cross-modal search can be defined as a set of retrieval applications that try to effectively find relevant documents of modality A by querying with documents from modality B.
+ - Multi-modal search can be defined as a set of retrieval applications that try to effectively project documents of different modalities into a
+ common embedding space, and find relevant documents with respect to the fusion of multiple modalities
+ 
+The main difference between these two search modes is that for cross-modal, there is a direct mapping between a single document or chunk and a
+vector in embedding space, while for MultiModal this does not hold true, since 2 or more documents might be combined into a single vector.
+ 
+This unlocks a lot of powerful patterns and makes Jina fully flexible and agnostic to what can be searched.
+ 
+ - It allows to look for images by giving corresponding caption descriptions (https://github.com/jina-ai/examples/tree/master/cross-modal-search)
```
Comment (yuanbit) [inline docs/chapters/cross-multi-modality/index.md:25]: [Image Search using Captions](https://github.com/jina-ai/examples/tree/master/cross-modal-search)
[Image Search using Images with Modification Text](https://github.com/jina-ai/examples/tree/master/multimodal-search-tirg)

File: README.md (hunk 9 above)
Comment (yuanbit) [inline README.md:399]: "composed by" should be "composed of"

"more than one chunks" should be "more than one chunk"

File: docs/chapters/cross-multi-modality/index.md [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+Jina does not only offer the possibility to work with any data, but also allows you to run cross- and multi-modal search flows. 
+To better understand what this implies we first need to understand the concept of modality. 
+
+One may think that different modalities correspond to different kinds of data (images and text in this case).
+However, this is not accurate. For example, one can do cross-modal search by searching images from different points of view, 
+or searching for matching titles for given paragraph text.
+Therefore, one can consider that a modality is related to a given data distribution from which input may come. 
+
+For this reason, and to have first-class support for cross and multi-modal search, Jina offers modality as an attribute in its Document primitive type.
+Now that we are agreed on the concept of modality, we can describe cross-modal and multi-modal search.
```
Comment (yuanbit) [inline docs/chapters/cross-multi-modality/index.md:14]: "Now that we are agreed on..." should be "Now that we agreed on ..."

File: docs/chapters/cross-multi-modality/index.md [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+ 
+ - Cross-modal search can be defined as a set of retrieval applications that try to effectively find relevant documents of modality A by querying with documents from modality B.
+ - Multi-modal search can be defined as a set of retrieval applications that try to effectively project documents of different modalities into a
+ common embedding space, and find relevant documents with respect to the fusion of multiple modalities
+ 
+The main difference between these two search modes is that for cross-modal, there is a direct mapping between a single document or chunk and a
+vector in embedding space, while for MultiModal this does not hold true, since 2 or more documents might be combined into a single vector.
```
Comment (yuanbit) [inline docs/chapters/cross-multi-modality/index.md:21]: "Now that we are agreed on..." should be "Now that we agreed on ..."

Comment (yuanbit) [reply docs/chapters/cross-multi-modality/index.md:21]: "2 or more..." should be "two or more..."

**Review by yuanbit — COMMENTED** [review]
I made some comments to reduce the amount of text of the MultimodalDocument section in the readme and the cross-multi-modality documentation.

I think the MultimodalDocument section is a bit out of place in the readme. I wonder if it would better serve its purpose in the documentation?

File: README.md (hunk 9 above)
Comment (alexcg1) [reply README.md:399]: @yuanbit i think at this point we've used the word Document so much that people should have a feel for it, especially if they're reading advanced stuff like multi-modal search

Other notes:
- `document` should be `Document` (bc it refers to a Jina-specific concept, not a document in general)
- `chunk` should be `Chunk` (same reason)
- `more than one` -> `multiple` (easier to read, and relates to the "multi" part)

Also, first time we mention each, let's link to the relevant section of Jina 101 that describes them. That gets around folks not understanding our usage of Document

File: docs/chapters/cross-multi-modality/index.md (hunk 13 above)
Comment (alexcg1) [reply docs/chapters/cross-multi-modality/index.md:3]: "Jina is a data type-agnostic framework, **letting you** work with any type of data and run cross- and multi-modal search **Flows**."

Emphasis to show my adjustments

File: docs/chapters/cross-multi-modality/index.md (hunk 14 above)
Comment (alexcg1) [reply docs/chapters/cross-multi-modality/index.md:9]: Let's keep the wording simple and second-person style

- `one` sounds like we're royalty. `you` or `we` is much more direct and commonly used irl
- `points of view` -> `perspectives` (I saw points of view at first and thought opinions)

File: docs/chapters/cross-multi-modality/index.md (hunk 17 above)
Comment (alexcg1) [reply docs/chapters/cross-multi-modality/index.md:21]: `2 or more` -> multiple

File: README.md (hunk 10 above)
Comment (alexcg1) [reply README.md:431]: Use Title Case for headings: https://titlecase.com/

File: README.md (hunk 12 above)
Comment (alexcg1) [reply README.md:456]: Agreed! I think we need to be very careful about overloading the readme. Better to put a general overview and then a link to more information

File: README.md (hunk 9 above)
Comment (alexcg1) [reply README.md:399]: multiple

File: docs/chapters/cross-multi-modality/index.md [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+ 
+This unlocks a lot of powerful patterns and makes Jina fully flexible and agnostic to what can be searched.
+ 
+ - It allows to look for images by giving corresponding caption descriptions (https://github.com/jina-ai/examples/tree/master/cross-modal-search)
+ - It allows to merge visual and textual information in a multi-modal way to look for images with transformation descriptions ( - It allows to look for images by giving corresponding caption descriptions (https://github.com/jina-ai/examples/tree/master/multi-modal-search-tirg)
```
Comment (alexcg1) [inline docs/chapters/cross-multi-modality/index.md:26]: Redo text: `It lets you merge visual and textual information in a multi-modal way search images using transformation descriptions (in this case, corresponding caption descriptions (https://github.com/jina-ai/examples/tree/master/multi-modal-search-tirg)`

Comment (alexcg1) [inline docs/chapters/cross-multi-modality/index.md:26]: - URL link is dead
- URL isn't attached to any text (I was expecting some markdown link syntax)

**Review by alexcg1 — CHANGES_REQUESTED** [review]
Nice work overall, but see my comments for small adjustments :smile: 

Main things to change throughout (and to bear in mind for future docs):

- Title case for headings
- Use `you` and `we` instead of `one`
- Keep vocabulary and structure as simple as we can - helps non-native speakers read more easily
- Capitalize Jina concepts, e.g. `Document`, `Chunk`, `Flow`

File: README.md (hunk 11 above)
Comment (JoanFM) [reply README.md:440]: I think though, it must be clear that the encoding is a merged one. How can this be clear and yet easier?

**Comment by JoanFM** [comment]
> I made some comments to reduce the amount of text of the MultimodalDocument section in the readme and the cross-multi-modality documentation.
> 
> I think the MultimodalDocument section is a bit out of place in the readme. I wonder if it would better serve its purpose in the documentation?

I may agree, @hanxiao should we keep it in the README?

**Comment by JoanFM** [comment]
Thanks for the great review @yuanbit @alexcg1 .

I applied almost every suggestion. I would agree it may feel out of place in the README.md but the idea was to have it highlighted specifically in the README.md.

**Review by nan-wang — APPROVED** [review]

**Comment by hanxiao** [comment]
im taking this PR now, will do minor change and merge in an hour
